sql3/planner/opcopy.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	sqlResponse := &pilosa.WireQueryResponse{}
+	err = sqlResponse.UnmarshalJSONTyped([]byte(body), true)
+	if err != nil {
+		return nil, sql3.NewErrInternalf("error executing remotely: %s", err.Error())
+	}
+
+	if len(sqlResponse.Error) > 0 {
+		return nil, sql3.NewErrInternalf("error executing remotely: %s", sqlResponse.Error)
+	}
+
+	return sqlResponse, nil
+}
+
+func (i *remoteCopyIterator) Next(ctx context.Context) (types.Row, error) {
+	if i.hasStarted == nil {
+		// execute the ddl to create the table
+		_, err := i.remoteExec(ctx, i.ddl)
+		if err != nil {
+			return nil, err
+		}
+
+		// build bulk insert statement
+		var buf bytes.Buffer
+		buf.WriteString("bulk insert into ")
+		fmt.Fprintf(&buf, "%s", i.targetTableName)
+		buf.WriteString(" (")
+
+		for i, s := range i.copySchema {
+			if i > 0 {
+				buf.WriteString(", ")
+			}
+			fmt.Fprintf(&buf, "%s", s.ColumnName)
+		}
+		buf.WriteString(") map (")
+		for i, s := range i.copySchema {
+			if i > 0 {
+				buf.WriteString(", ")
+			}
+			fmt.Fprintf(&buf, "'$._%d' %s", i, s.Type.TypeDescription())
+		}
+		buf.WriteString(") from x'")
+		header := buf.String()
+
+		batchCount := 0
+		var batchBuf bytes.Buffer
+
+		for {
+			// get a source row
+			row, err := i.child.Next(ctx)
+			if err != nil {
+				if err == types.ErrNoMoreRows {
+					break
+				}
+				return nil, err
+			}
+
+			// add it to target batch
+			var rowBuf bytes.Buffer
+			rowBuf.WriteString("{")
+			for i, s := range i.copySchema {
+				if i > 0 {
+					rowBuf.WriteString(",")
+				}
+				fmt.Fprintf(&rowBuf, `"_%d":`, i)
+
+				if row[i] == nil {
+					rowBuf.WriteString("null")
+					continue
+				}
+
+				switch ty := s.Type.(type) {
+				case *parser.DataTypeID, *parser.DataTypeInt:
+					val, ok := row[i].(int64)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected type '%T'", row[i])
+					}
+					fmt.Fprintf(&rowBuf, "%d", val)
+
+				case *parser.DataTypeString:
+					val, ok := row[i].(string)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected type '%T'", row[i])
+					}
+					// escape single quotes
+					val = strings.ReplaceAll(val, `'`, `''`)
+					// and double quotes
+					val = strings.ReplaceAll(val, `"`, `\"`)
+					// and line feeds
+					if strings.Contains(val, "\n") {
+						val = strings.ReplaceAll(val, "\n", "\\n")
+					}
+					fmt.Fprintf(&rowBuf, `"%s"`, val)
+
+				case *parser.DataTypeBool:
+					val, ok := row[i].(bool)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected type '%T'", row[i])
+					}
+					fmt.Fprintf(&rowBuf, "%v", val)
+
+				case *parser.DataTypeTimestamp:
+					val, ok := row[i].(time.Time)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected type '%T'", row[i])
+					}
+					fmt.Fprintf(&rowBuf, `"%s"`, val.Format(time.RFC3339Nano))
+
+				case *parser.DataTypeStringSet:
+					val, ok := row[i].([]string)
+					if !ok {
+						return nil, sql3.NewErrInternalf("unexpected type '%T'", row[i])
+					}
```

Comment:
i believe in both this case and DataTypeIDSet, we should check for/support NULL values.

I *think* that null values should always show up as a slice that is nil, as contrasted with a zero-length slice, but I slightly distrust that, and think we should in full generality probably accept a nil in `row[i]` as a NULL for the inserted value. I believe in inserts we can write a literal NULL in those slots.

Reply:
Do we know for a fact we can have a null value in a set?